Rework vehicle movement on rails

olanti-p opened the pull request:
#### Summary
SUMMARY: Bugfixes "Made vehicle movement on rails more robust"

#### Purpose of change
In order to have actual usable railroads on game map, vehicle movement on rails has to be fixed first.
Currently it has the following problems:
1. It is unclear how rail wheels interact with rails. Going off of [this reddit post](https://old.reddit.com/r/cataclysmdda/comments/nuq4ib/all_aboard_the_subterranean_death_train/h0yvnr8/), it took a lot of experimentation to make a player-built vehicle somewhat reliably use rails (turn when needed, not derail).
2. It is unclear under what circumstances the vehicle will turn on rails, or whether it will turn at all. Subway t-junctions have this weird setup where one of their rail segments is missing for some reason. Is this made this way to forbid 2-railed vehicles (like draisines) from turning? Or does it just de-prioritize autoturn algorithm from taking a turn, and not following the straight path?
3. It is unclear under what circumstances the vehicle will derail. Sometimes it derails on t-junctions, sometimes it does not derail on t-junctions, sometimes it derails on regular turns, sometimes it works fine and takes a turn. Trying to manually take a left turn on a right turn is a sure way of getting wrecked.
4. It is very hard to get vehicle back on tracks. Not only you have to perfectly line it up: even if you do succeed with that, there's a chance the vehicle will "snap" to the rails in a semi-rotated state, so on each game turn it will simply alternate between 2 semi-rotated states instead of going straight. Or snap improperly, e.g. draisine having its right wheels snapped on the left rail of the track, with left wheels dangling outside and scraping walls.
5. Sometimes vehicles change tracks for no reason whatsoever (https://github.com/CleverRaven/Cataclysm-DDA/pull/30232#issuecomment-492166777).
6. The list of accepted rail terrains is hardcoded.

#### Describe the solution
Remove the old movement algorithm that kept track of individual rail wheels and their positions on rail tracks, and replace it with a more abstract one. The new algorithm first builds a "rail profile" of the vehicle, which describes where the vehicle would expect rails to be on its y axis relative to its pivot point in its own coordinate space. For example, for 2-seated draisine the rail profile would look like:
```
      +x
   ..:...:..      - and |    draisine frame
   ..:...:..         0         railwheels
   ..0---0..         p        pivot point
-y ..|---|.. +y      .       ground
   ..0-p-0..         :     expected rail
   ..:...:..   
   ..:...:..     rail profile: [ -2, 2 ]
      -x
```
Then, during movement, the algorithm tries to match surrounding terrain to expected rail profile by searching for 3 tiles with `RAIL` terrain flag in a row at the proper offset and direction from pivot point. There following options are scanned:
1. Direction of vehicle movement, to check whether terrain in front of us contains valid rails
2. Direction of vehicle movement, but with additional +1/-1 y offset, to check for "strafing" rails
3. Direction of vehicle movement +45/-45 degrees, to check for turns

Once the algorithm has determined which movement modes are available, it selects one of them based on whether (and where) the player is trying to turn the vehicle, and produces the result.

If vehicle is derailed, the algorithm checks for rails in nearby possible rail direction (a multiple of 45 degrees). If there are rails there, the vehicle facing will be snapped to that direction, which will case it to underail.

For the rest of the codebase, vehicle is considered to be on rails if it has valid direction (a multiple of 45 degrees) and a valid segment of rails in front or behind of its pivot point (when the vehicle is moving through a turn, or "strafing", one of the two is guaranteed to be true).

Strafing (or "shifting" in code) movement allows shifting vehicle position by a few tiles while moving without actually changing its facing, which helps in tight spaces where 45 degree turn is either ugly or impossible to make.

#### TODO list
- [x] Handle z-level movement
- [x] Make it so all rail wheels have low friction when vehicle is considered to be on rails, regardless of whether rail wheels are actually on rail tiles.
- [x] Make it impossible to pull off u-turns using handbreak
- [x] Reduce max angle at which the vehicle snaps back onto rails
- [x] Sprinkle code with documentation

#### Describe alternatives you've considered
- Have the vehicle keep some kind of state to keep track of how it's moving on rails
  - This is rather bug prone, since it'll have to be kept in sync with movement from other sources (e.g. collisions) and map changes.
- Add a special "rail fork" terrain flag that allows a rail to be marked as fork, so the vehicle can switch tracks only on it
  - This might be necessary in the future, but for now a single `RAIL` terrain flag is sufficient.
- Implement actual smooth turns (with 15 and 30 degree steps) on rails
  - Vehicles turned to 15 or 30 degree don't cleanly map to map tiles during movement, which is a pain. It could be possible with internal "rail movement state", but as mentioned above, that'd be sub-optimal too.
  - Strafing is a workaround for this: instead of turning 15 degrees and shifting position by 1 every x tiles, you can do the same shift without relying on vehicle skew to match terrain. Additionally, with right combination of strafing rails it's possible to make a somewhat smooth turn: 
<details>
<summary>picture</summary>

![image](https://user-images.githubusercontent.com/60584843/174495625-d0656642-69eb-4f32-a1d2-72286ef5d7da.png)
</details>

#### Testing
Unit tests, also manual in another branch (too much code to post there).

#### Additional context
Well, this took a while.
This PR does not fix the problem of ramming subway walls on turns, but since vehicles would be able to take smooth turns, this should be fixable with proper mapgen.

Zireael07 commented:
Like it! Once it's done, probably a candidate for backporting to CDDA?

AngelicosPhosphoros commented on tests/map_setup_helpers_test.cpp:
```diff
@@ -0,0 +1,266 @@
+    if( candidates.empty() ) {
+        return cata::nullopt;
+    } else {
+        return candidates[0];
```
`return candidates.front();`?

AngelicosPhosphoros commented on tests/map_setup_helpers_test.cpp:
```diff
@@ -0,0 +1,266 @@
+
+canvas canvas::rotated( int turns ) const
+{
+    point new_size = size_cache.rotate( turns ).abs();
```
`const`

AngelicosPhosphoros commented on tests/map_setup_helpers_test.cpp:
```diff
@@ -0,0 +1,266 @@
+            const std::string &exp = l->entry( expected.get( p ) );
+            std::string got = getter( p );
+            if( exp != got ) {
+                fails.push_back( { p, exp, got } );
```
`std::move(got)`?

AngelicosPhosphoros commented on tests/stringmaker.h:
```diff
@@ -8,12 +8,23 @@
 #include "item.h"
+#include "units_angle.h"
+
+#include <utility>
```
It is more common to include STL libraries before other includes.

olanti-p commented on src/vehicle.h:
```diff
@@ -1779,6 +1763,8 @@ class vehicle
         std::vector<int> speciality;
         std::vector<int> floating;         // List of parts that provide buoyancy to boats
 
+        std::vector<int> rail_profile; // Rail profile of the vehicle
```
Yeah, that was a temporary thing. I've ended up copying over the explanation from the PR text.

olanti-p commented on src/vehicle_move.cpp:
```diff
@@ -1823,3 +1651,206 @@ units::angle map::shake_vehicle( vehicle &veh, const int velocity_before,
+    point veh_plus_y_vec,
+    point scan_vec )
+{
+    map &here = get_map();
```
True. Moved it to argument as well.

olanti-p commented on src/vehicle_move.cpp:
```diff
@@ -1823,3 +1651,206 @@ units::angle map::shake_vehicle( vehicle &veh, const int velocity_before,
+    return false;
+}
+
+inline rail_processing_result make_none()
```
MSVC doesn't catch these... Done.

olanti-p commented:
Ok, made some changes:
* `const`ified a bunch of thing, made other tweaks suggested in @AngelicosPhosphoros's review
* added a few more tests
* found out the existing code already perfectly supports rail ramp transitions (lol), added tests just in case
* fixed a bug with getting back on tracks
* made it impossible to derail by changing direction & pulling handbreak
* made it so vehicles on rails don't lose speed on turns/forks (an abstraction for how they're supposed to turn gradually and have all their wheels on rails at all times)

Now waiting for further reviews and clang-tidy's verdict.
All rail movement tests take >200 seconds on my machine, so it'll probably be better to disable most of them by default before merging.

Coolthulhu commented:
Nice!

A weird interaction:
* Spawn the 6 wheel rail cart thing
* Drive over standard double rail so that one pair of wheels is over the rail
* Turn
* Velocity suddenly drops

Might be related to snapping-on behavior.
It's not necessarily wrong, just different - other vehicles keep all their velocity even when turning at impossible angles.
